src/yz_solrq_helper.erl

```diff
@@ -395,3 +505,50 @@ terminate(_Reason, _State) ->
 
 code_change(_OldVsn, State, _Extra) ->
     {ok, State}.
+
+%% @doc Entries is [{BKey, Obj, Reason, P, ShortPL, Hash}]
+%% These are in chronological order - we need to get the latest for each BKey
+get_latest_entries_for_keys(Entries) ->
+    EntryDict = lists:foldl(fun entry_given_previous_entries/2, dict:new(), Entries),
+    dict:fold(fun add_value_to_list/3, [], EntryDict).
+
+add_value_to_list(_Key, Value, List) ->
+    [Value | List].
+
+entry_given_previous_entries(Entry, PreviousEntries) ->
```

Comment:
This would be simpler if we leverage `dict:update/4` along these lines:
```
entry_given_previous_entries(Entry, PreviousEntries) ->
    NewReason = get_reason_from_entry(Entry),
    BKey = get_bkey_from_entry(Entry),
    dict:update(BKey,
                        fun (PreviousEntry) ->
                               PreviousReason = get_reason_from_entry(PreviousEntry),
                               case choose_entry(NewReason, PreviousReason) of
                                   new -> Entry;
                                  _  -> PreviousEntry
                               end
                         end,
                         Entry,
                         PreviousEntries).
```

Having written this I would even say that we should let `choose_entry/2` take Entries as argument and rename the existing `choose_entry/2` to `compare_entry_reasons/2`

```
choose_entry(Entry, PreviousEntry) ->
     NewReason = get_reason_from_entry(Entry),
     PreviousReason = get_reason_from_entry(PreviousEntry),
     case compare_entry_reasons(NewReason, PreviousReason) of
                                   new -> Entry;
                                  _  -> PreviousEntry
                               end.
``` 
And then we get:
```
entry_given_previous_entries(Entry, PreviousEntries) ->
    BKey = get_bkey_from_entry(Entry),
    dict:update(BKey,
                        fun (PreviousEntry) ->
                              choose_entry(Entry, PreviousEntry)
                        end,
                        Entry,
                        PreviousEntries).
```